File: web_ui/src/drags.js
// A drag is a run of trial edits between begin_drag and end_drag, which the kernel records
// as one step or none. Ported from the add-on's viewport/live.py.
import { op, OpError } from "./api.js";
import { useStore } from "./store.js";

export class LiveDrag {
  // make(value) -> [op, args] creates the feature; edit(value) -> args edits it after
  constructor({ label, make, edit, format = (v) => `${v.toFixed(2)} mm` }) {
    Object.assign(this, { label, make, edit, format });
    this.feature = null; this.value = null; this.good = null; this.refused = null;
    this.typed = ""; this.busy = false; this.pending = null; this.done = false;
  }

  async begin() {
    await op("begin_drag");
    useStore.setState({ drag: { label: this.label, text: "", refused: false, x: 0, y: 0 } });
    this._onKey = (e) => this.key(e);
    window.addEventListener("keydown", this._onKey, true);
    return this;
  }

  // the badge beside the cursor: what the drag is at, and whether it built
  say(text, refused = false, at = null) {
    useStore.setState((s) => ({ drag: { ...(s.drag || {}), label: this.label, text, refused, ...(at ? { x: at.x, y: at.y } : {}) } }));
  }

  preview(value, at = null) {
    if (this.done) return;
    this.value = value;
    this.say(this.format(value), false, at);
    this.pending = value;
    if (!this.busy) this._run();
  }

  async _run() {
    this.busy = true;
    while (this.pending !== null && !this.done) {
      const value = this.pending; this.pending = null;
      try {
        let out;
        if (this.feature === null) out = await op(...this.make(value));
        else {
          // edit answers the arguments to change, or [op, args] for an operation of its own
          const change = this.edit(value, this.feature);
          out = Array.isArray(change) ? await op(...change) : await op("edit_feature", { feature_id: this.feature, args: change });
        }
        if (this.feature === null) this.feature = out.feature;
        this.good = value; this.refused = null;
        await useStore.getState().shown(out);
        if (!this.done) this.say(this.format(value), false);
      } catch (err) {
        // the part on screen is still the last one that built: say so in the badge
        this.refused = value;
        this.say(err instanceof OpError ? err.message : String(err), true);
      }
    }
    this.busy = false;
  }

  key(e) {
    if (e.key === "Escape") { e.preventDefault(); this.cancel(); return; }
    if (e.key === "Enter") { e.preventDefault(); this.finish(); return; }
    if (/^[0-9.\-]$/.test(e.key)) { this.typed += e.key; e.preventDefault(); }
    else if (e.key === "Backspace") { this.typed = this.typed.slice(0, -1); e.preventDefault(); }
    else return;
    const typed = Number(this.typed);
    if (this.typed && Number.isFinite(typed)) { this.preview(typed); this.say(`${this.typed} mm (typed)`); }
  }

  _theValueThatBuilt() { return this.refused !== null && this.good !== null ? this.good : this.value; }

  async finish() {
    if (this.done) return;
    this.done = true; this.pending = null;
    window.removeEventListener("keydown", this._onKey, true);
    const value = this._theValueThatBuilt();
    const store = useStore.getState();
    if (value === null) { await this._drop(); return; }
    try {
      await op("reset_drag");
      await op(...this.make(value));
      await store.shown(await op("end_drag", { keep: true }));
      store.say(`${this.label}: ${this.format(value)}${this.refused !== null ? " (the shape would not take any more)" : ""}`);
    } catch (err) {
      try { await store.shown(await op("end_drag", { keep: false })); } catch { /* already gone */ }
      store.say(err instanceof OpError ? `${err.kind}: ${err.message}` : String(err), true);
    }
    useStore.setState({ drag: null });
  }

  async cancel() {
    if (this.done) return;
    this.done = true; this.pending = null;
    window.removeEventListener("keydown", this._onKey, true);
    await this._drop();
    useStore.getState().say("cancelled");
  }

  async _drop() {
    try { await useStore.getState().shown(await op("end_drag", { keep: false })); } catch { /* nothing to drop */ }
    useStore.setState({ drag: null });
  }
}

// Snap targets, built once per drag from describe_faces: parallel planes, and points on the face.
export class Snapper {
  static PIXELS = 9;
  constructor(faces, frame, own) {
    const n = frame.normal, o = frame.origin;
    const dot = (a, b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
    this.planes = [];
    for (const f of faces) {
      if (f.shape !== "plane" || f.name === own || !f.normal) continue;
      if (Math.abs(Math.abs(dot(n, f.normal)) - 1) > 1e-3) continue;
      const d = dot([f.centre[0] - o[0], f.centre[1] - o[1], f.centre[2] - o[2]], n);
      if (Math.abs(d) > 1e-6) this.planes.push([d, f.name]);
    }
    this.points = [[[0, 0], "centre"]];
    for (const f of faces) {
      if (f.shape !== "cylinder" || !f.axis) continue;
      if (Math.abs(Math.abs(dot(n, f.axis)) - 1) > 1e-3) continue;
      this.points.push([toUV(f.centre, frame), f.name]);
    }
  }
  distance(raw, mmPerPixel) {
    let best = raw, label = null;
    for (const [d, name] of this.planes) {
      if (Math.abs(d - raw) <= Snapper.PIXELS * mmPerPixel && (label === null || Math.abs(d - raw) < Math.abs(best - raw))) { best = d; label = name; }
    }
    return [best, label];
  }
  uv(uv, mmPerPixel) {
    const reach = Snapper.PIXELS * mmPerPixel;
    for (const [[pu, pv], name] of this.points) if (Math.hypot(uv[0] - pu, uv[1] - pv) <= reach) return [[pu, pv], name];
    for (const [[pu, pv], name] of this.points) {
      if (Math.abs(uv[0] - pu) <= reach) return [[pu, uv[1]], `${name} (u)`];
      if (Math.abs(uv[1] - pv) <= reach) return [[uv[0], pv], `${name} (v)`];
    }
    return [uv, null];
  }
}

// a face frame: origin, outward normal, x_axis; (u, v) are millimetres in the face's own plane
export function toUV(point, frame) {
  const d = [point[0] - frame.origin[0], point[1] - frame.origin[1], point[2] - frame.origin[2]];
  const x = frame.x_axis, n = frame.normal;
  const y = [n[1] * x[2] - n[2] * x[1], n[2] * x[0] - n[0] * x[2], n[0] * x[1] - n[1] * x[0]];
  return [d[0] * x[0] + d[1] * x[1] + d[2] * x[2], d[0] * y[0] + d[1] * y[1] + d[2] * y[2]];
}
export function to3D(uv, frame) {
  const x = frame.x_axis, n = frame.normal;
  const y = [n[1] * x[2] - n[2] * x[1], n[2] * x[0] - n[0] * x[2], n[0] * x[1] - n[1] * x[0]];
  return [0, 1, 2].map((i) => frame.origin[i] + x[i] * uv[0] + y[i] * uv[1]);
}
export const snap = (value, step) => Math.round(value / step) * step;

File: native_app/drags.py
"""A drag is a run of trial edits between begin_drag and end_drag, which the kernel records
as one step or none. The same contract as the add-on's viewport/live.py and the page's drags.js."""
from __future__ import annotations

import math

from PySide6.QtCore import QObject, QTimer, Signal

from cadcore.service import web


def _takes_two(fn) -> bool:
    import inspect
    try:
        return len(inspect.signature(fn).parameters) >= 2
    except (TypeError, ValueError):
        return False


class LiveDrag(QObject):
    """make(value) -> (op, args) creates the feature; edit(value[, feature]) -> args edits it after,
    or (op, args) for an operation of its own."""

    changed = Signal()          # the badge: text, refused, value
    ended = Signal()

    def __init__(self, window, label: str, make, edit, fmt=None):
        super().__init__(window)
        self.window, self.label, self.make, self.edit = window, label, make, edit
        self.fmt = fmt or (lambda v: "%.2f mm" % v)
        self.feature = None
        self.value = self.good = self.refused = self.pending = None
        self.typed = ""
        self.text, self.is_refused, self.busy, self.done = "", False, False, False

    def begin(self) -> "LiveDrag":
        self.window.op("begin_drag")
        return self

    def say(self, text: str, refused: bool = False) -> None:
        self.text, self.is_refused = text, refused
        self.changed.emit()

    def preview(self, value: float) -> None:
        if self.done:
            return
        self.value = value
        self.say(self.fmt(value))
        self.pending = value
        if not self.busy:
            QTimer.singleShot(0, self._run)         # after this event: forty moves are a few rebuilds

    def _run(self) -> None:
        self.busy = True
        try:
            while self.pending is not None and not self.done:
                value, self.pending = self.pending, None
                try:
                    if self.feature is None:
                        op, args = self.make(value)
                        out = self.window.op(op, **args)
                        self.feature = out.get("feature")
                    else:
                        change = self.edit(value, self.feature) if _takes_two(self.edit) else self.edit(value)
                        if isinstance(change, tuple):            # an operation of its own
                            out = self.window.op(change[0], **change[1])
                        else:
                            out = self.window.op("edit_feature", feature_id=self.feature, args=change)
                    self.good, self.refused = value, None
                    self.window.shown(out)
                    if not self.done:
                        self.say(self.fmt(value))
                except web.CadError as exc:
                    # the part on screen is still the last one that built: say so in the badge
                    self.refused = value
                    self.say(exc.message, refused=True)
        finally:
            self.busy = False

    def key(self, text: str) -> bool:
        """A typed digit, a Backspace ("\\b"), Enter or Escape; whether it was taken."""
        if text == "escape":
            self.cancel()
            return True
        if text == "enter":
            self.finish()
            return True
        if text == "\b":
            self.typed = self.typed[:-1]
        elif text and text in "0123456789.-":
            self.typed += text
        else:
            return False
        try:
            typed = float(self.typed)
        except ValueError:
            return True
        self.preview(typed)
        self.say("%s mm (typed)" % self.typed)
        return True

    def _the_value_that_built(self):
        return self.good if self.refused is not None and self.good is not None else self.value

    def finish(self) -> None:
        if self.done:
            return
        self.done, self.pending = True, None
        value = self._the_value_that_built()
        window = self.window
        if value is None:
            self._drop()
            return
        try:
            window.op("reset_drag")
            op, args = self.make(value)
            window.op(op, **args)
            window.shown(window.op("end_drag", keep=True))
            window.say("%s: %s%s" % (self.label, self.fmt(value),
                                     " (the shape would not take any more)" if self.refused is not None else ""))
        except web.CadError as exc:
            try:
                window.shown(window.op("end_drag", keep=False))
            except web.CadError:
                pass
            window.say("%s: %s" % (exc.kind, exc.message), error=True)
        self.ended.emit()

    def cancel(self) -> None:
        if self.done:
            return
        self.done, self.pending = True, None
        self._drop()
        self.window.say("cancelled")

    def _drop(self) -> None:
        try:
            self.window.shown(self.window.op("end_drag", keep=False))
        except web.CadError:
            pass
        self.ended.emit()


class Snapper:
    """Snap targets, built once per drag from describe_faces: parallel planes, points on the face."""

    PIXELS = 9.0

    def __init__(self, faces: list, frame: dict, own: str):
        n, o = frame["normal"], frame["origin"]
        self.planes, self.points = [], [((0.0, 0.0), "centre")]
        for f in faces:
            if f.get("shape") == "plane" and f["name"] != own and f.get("normal"):
                if abs(abs(_dot(n, f["normal"])) - 1) > 1e-3:
                    continue
                d = _dot(_sub(f["centre"], o), n)
                if abs(d) > 1e-6:
                    self.planes.append((d, f["name"]))
            elif f.get("shape") == "cylinder" and f.get("axis"):
                if abs(abs(_dot(n, f["axis"])) - 1) > 1e-3:
                    continue
                self.points.append((to_uv(f["centre"], frame), f["name"]))

    def distance(self, raw: float, mm_per_pixel: float) -> tuple:
        best, label = raw, None
        for d, name in self.planes:
            if abs(d - raw) <= self.PIXELS * mm_per_pixel and (label is None or abs(d - raw) < abs(best - raw)):
                best, label = d, name
        return best, label

    def uv(self, uv: tuple, mm_per_pixel: float) -> tuple:
        reach = self.PIXELS * mm_per_pixel
        for (pu, pv), name in self.points:
            if math.dist(uv, (pu, pv)) <= reach:
                return (pu, pv), name
        for (pu, pv), name in self.points:
            if abs(uv[0] - pu) <= reach:
                return (pu, uv[1]), name + " (u)"
            if abs(uv[1] - pv) <= reach:
                return (uv[0], pv), name + " (v)"
        return uv, None


# a face frame: origin, outward normal, x_axis; (u, v) are millimetres in the face's own plane
def to_uv(point, frame: dict) -> tuple:
    d = _sub(point, frame["origin"])
    x, n = frame["x_axis"], frame["normal"]
    y = _cross(n, x)
    return (_dot(d, x), _dot(d, y))


def to_3d(uv, frame: dict) -> list:
    x, n = frame["x_axis"], frame["normal"]
    y = _cross(n, x)
    return [frame["origin"][i] + x[i] * uv[0] + y[i] * uv[1] for i in range(3)]


def snap(value: float, step: float) -> float:
    return round(value / step) * step


def _dot(a, b) -> float:
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def _sub(a, b) -> list:
    return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]


def _cross(a, b) -> list:
    return [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]]
